ros2topic/ros2topic/verb/pub.py

```diff
@@ -64,13 +64,23 @@ def publisher(message_type, topic_name, values):
     pub = node.create_publisher(msg_module, topic_name)
 
     msg = msg_module()
-    for field_name, field_value in values_dictionary.items():
-        field_type = type(getattr(msg, field_name))
-        setattr(msg, field_name, field_type(field_value))
+    set_msg_fields(msg, values_dictionary)
 
     print('publisher: beginning loop')
     while rclpy.ok():
         pub.publish(msg)
         print('publishing %r\n' % msg)
         time.sleep(1)
     rclpy.shutdown()
+
+
+def set_msg_fields(msg, values):
+    for field_name, field_value in values.items():
+        print(field_name, field_value)
+        field_type = type(getattr(msg, field_name))
+        try:
+            value = field_type(field_value)
+        except TypeError:
+            value = field_type()
+            set_msg_fields(value, field_value)
+        setattr(msg, field_name, value)
```

Comment:
I recommend catching exceptions here and saying which field failed, otherwise here's an example of giving 35 elements to Covariance instead of 36:

```
$ ros2 topic pub asdf geometry_msgs/PoseWithCovariance "pose:
  position: {x: 0.0, y: 0.0, z: 0.0}
  orientation: {x: 0.0, y: 0.0, z: 0.0, w: 0.0}
covariance: [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
  0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
  0.0, 0.0, 0.0, 0.0, 0.0]" 
covariance [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
Traceback (most recent call last):
  File "/home/dhood/ros2_ws/install_isolated/ros2cli/bin/ros2", line 11, in <module>
    load_entry_point('ros2cli==0.0.0', 'console_scripts', 'ros2')()
  File "/home/dhood/ros2_ws/install_isolated/ros2cli/lib/python3.5/site-packages/ros2cli/cli.py", line 64, in main
    rc = extension.main(parser=parser, args=args)
  File "/home/dhood/ros2_ws/install_isolated/ros2topic/lib/python3.5/site-packages/ros2topic/command/topic.py", line 38, in main
    return extension.main(args=args)
  File "/home/dhood/ros2_ws/install_isolated/ros2topic/lib/python3.5/site-packages/ros2topic/verb/pub.py", line 43, in main
    return main(args)
  File "/home/dhood/ros2_ws/install_isolated/ros2topic/lib/python3.5/site-packages/ros2topic/verb/pub.py", line 47, in main
    publisher(args.message_type, args.topic_name, args.values)
  File "/home/dhood/ros2_ws/install_isolated/ros2topic/lib/python3.5/site-packages/ros2topic/verb/pub.py", line 67, in publisher
    set_msg_fields(msg, values_dictionary)
  File "/home/dhood/ros2_ws/install_isolated/ros2topic/lib/python3.5/site-packages/ros2topic/verb/pub.py", line 86, in set_msg_fields
    setattr(msg, field_name, value)
  File "/home/dhood/ros2_ws/install_isolated/geometry_msgs/lib/python3.5/site-packages/geometry_msgs/msg/_pose_with_covariance.py", line 102, in covariance
    True)
AssertionError
```